### PR TITLE
Labs 67 mysql user toggle service

### PR DESCRIPTION
**Pull request by vokoblin**
Makes and satisfies all the tests for MysqlUserToggleService.
Modifies the tables.sql to add user_activated_toggle table.

### CODE REVIEW

**Comment by vokoblin** [comment]
Waiting for https://github.com/clearbooks/labs/pull/28 to be merged after which this will be re-factored
.

**Comment by vokoblin** [comment]
More re-factoring needs to be done!

**Comment by viktorcb** [comment]
You have a duplication

**Comment by vokoblin** [comment]
WOOOOOOHOOOOOO FINALY....

File: src/User/MysqlToggleStatusModifierService.php [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+            return $this->tryInsertElseUpdateToggleToActiveState( $toggleIdentifier, $userOrGroupIdentifier, $isGroup );
+        } else if ( $this->toggleStatusInactive( $toggleStatus ) ) {
+            return $this->tryInsertElseUpdateToggleToInactiveState( $toggleIdentifier, $userOrGroupIdentifier,
+                $isGroup );
+        } else if ( $this->toggleStatusUnset( $toggleStatus ) ) {
+            return $this->unsetToggle( $toggleIdentifier, $userOrGroupIdentifier, $isGroup );
+        } else {
+            return false;
+        }
+    }
```
Comment (MeliaBurnley) [inline src/User/MysqlToggleStatusModifierService.php:283]: just for in the future, you can have the return true after the closing brace of the else: 

```
} else { 
    return false;
}
return true;
```

saves having all the returns inside the if statements, but that's not important i dont think. But seeing as they all return true it saves having lots of return points

File: test/User/MysqlToggleStatusModifierServiceTest.php [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+     */
+    private function getUserActivatedToggleEntry( $toggle_id, $user_id )
+    {
+        $data = $this->connection->fetchAssoc( 'SELECT * FROM `user_activated_toggle` WHERE toggle_id = ? AND user_id = ?',
+            [ $toggle_id, $user_id ] );
+        if ( empty( $data ) ) {
+            return [ ];
+        }
+        $entery = [ 1 => $data[ 'toggle_id' ], 2 => $data[ 'user_id' ], 3 => $data[ 'is_active' ] ];
+        return $entery;
```
Comment (MeliaBurnley) [inline test/User/MysqlToggleStatusModifierServiceTest.php]: entry*

File: test/User/MysqlToggleStatusModifierServiceTest.php [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+     */
+    private function getGroupActivatedToggleEntry( $toggle_id, $group_id )
+    {
+        $data = $this->connection->fetchAssoc( 'SELECT * FROM `group_activated_toggle` WHERE toggle_id = ? AND group_id = ?',
+            [ $toggle_id, $group_id ] );
+        if ( empty( $data ) ) {
+            return [ ];
+        }
+        $entery = [ 1 => $data[ 'toggle_id' ], 2 => $data[ 'group_id' ], 3 => $data[ 'active' ] ];
+        return $entery;
```
Comment (MeliaBurnley) [inline test/User/MysqlToggleStatusModifierServiceTest.php]: entry*

File: src/User/MysqlToggleStatusModifierService.php (hunk 1 above)
Comment (vokoblin) [inline src/User/MysqlToggleStatusModifierService.php:283]: Thanks, :dancer: I know, preference I guess don't like letting it execute when it needs to terminate there.

**Comment by MeliaBurnley** [comment]
That was long, but this looks good now :+1:

**Pull request merged**